### PR TITLE
adding typescript definition

### PR DESCRIPTION
**Pull request by CBaptiste**
Hello,

Let's integrate the whole typescript definition of algoliasearch !

### CODE REVIEW

**Comment by CBaptiste** [comment]
I need to double check all def and properties but I think it is ok.
what do you think @vvo ?

**Comment by CBaptiste** [comment]
👍 
ts

File: typings/algoliasearch.d.ts [hunk 1]
```diff
@@ -0,0 +1,1069 @@
...
+    alternativesAsExact?: any;
+    /**
+     * The name of the attribute used for the Distinct feature
+     * default: null
+     */
+    attributeForDistinct?: string;
+    /**
+     * If set to 1, enables the distinct feature, disabled by default, if the attributeForDistinct index setting is set.
+     */
+    distinct?: number;
```
Comment (redox) [inline typings/algoliasearch.d.ts]: `distinct` can also be `true` or `false`. Not sure how to handle it :/

Comment (redox) [inline typings/algoliasearch.d.ts]: Looks like we need to use `any`.

Comment (CBaptiste) [inline typings/algoliasearch.d.ts]: yes yes, for dynamically typed param, we ll use any !
thanks for spotting it out !

File: typings/algoliasearch.d.ts [hunk 2]
```diff
@@ -0,0 +1,1252 @@
...
+     */
+    initIndex(name:string):IAlgoliaIndex;
+    /**
+     * Initialization of the client
+     * @param name of the index
+     * @param params options related to the client initialization
+     * return algolia index object
+     * https://github.com/algolia/algoliasearch-client-js#init-index---initindex
+     */
+    initIndex(name:string, params:IClientOptions):IAlgoliaIndex;
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: I think there's no way to call initIndex(name, params), can you link the code?

File: typings/algoliasearch.d.ts [hunk 3]
```diff
@@ -0,0 +1,1252 @@
...
+     */
+    search(queries:{ indexName:string, query:string, options:IAlgoliaQueryParameters }, params:any, cb:Function);
+    /**
+     * Query on multiple index
+     * @param queries index name, query and query parameters
+     * @param params query parameter
+     * @param cb callback(err, res)
+     * https://github.com/algolia/algoliasearch-client-js#multiple-queries---multiplequeries
+     */
+    search(queries:{ indexName:string, query:string, options:IAlgoliaQueryParameters }, params:any, cb:Function, err:Function);
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: There's no err:Function (cb will be called with (err, res))

File: typings/algoliasearch.d.ts [hunk 4]
```diff
@@ -0,0 +1,1252 @@
...
+     * @param queries index name, query and query parameters
+     * @param cb callback(err, res)
+     * https://github.com/algolia/algoliasearch-client-js#multiple-queries---multiplequeries
+     */
+    search(queries:{ indexName:string, query:string, options:IAlgoliaQueryParameters }, params:any, cb:Function);
+    /**
+     * Query on multiple index
+     * @param queries index name, query and query parameters
+     * @param params query parameter
+     * @param cb callback(err, res)
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: Can we add that every (asynchronous) method returns a Promise if not callback passed?

File: typings/algoliasearch.d.ts [hunk 5]
```diff
@@ -0,0 +1,1252 @@
...
+     * https://github.com/algolia/algoliasearch-client-js#add-objects---addobjects
+     */
+    addObjects(objects:Object[], cb:Function);
+    /**
+     * Add or replace a specific object
+     * @param object
+     * @param cb(err, res)
+     * https://github.com/algolia/algoliasearch-client-js#update-objects---saveobjects
+     */
+    saveObject(object:Object, cb:Function);
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: cb is always optionnal, you get a Promise when no cb.

File: typings/algoliasearch.d.ts [hunk 6]
```diff
@@ -0,0 +1,1252 @@
...
+     */
+    setSettings(settings:IAlgoliaIndexSettings, cb:Function);
+    /**
+     * Search in an index
+     * @param query string
+     * @param params query parameters
+     * return Promise
+     * https://github.com/algolia/algoliasearch-client-js#search-in-an-index---search
+     */
+    search(query:string, params:IAlgoliaQueryParameters):Promise<IAlgoliaResponse>;
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: I think for search we should only advise to use the form:

```js
search({params}, [cb]}
```

query, params is not deprecated but there's no real advantage of using it. Makes the API simpler

File: typings/algoliasearch.d.ts [hunk 7]
```diff
@@ -0,0 +1,1252 @@
...
+    search(query:string, params:IAlgoliaQueryParameters, cb:Function);
+    /**
+     * Search in an index
+     * @param query query string
+     * @param params query parameter
+     * @param cb(err, res)
+     * @param err() error callback
+     * https://github.com/algolia/algoliasearch-client-js#search-in-an-index---search
+     */
+    search(query:string, params:IAlgoliaQueryParameters, cb:Function, err:Function);
```
Comment (vvo) [inline typings/algoliasearch.d.ts]: no err function

**Comment by CBaptiste** [comment]
will be done on definitelytyped instead

**Pull request closed without merging**